File: simplicadet.app/app/src/main/java/com/cadetia/simplicadet/activities/PathSelector.java
package com.cadetia.simplicadet.activities;

import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;

import com.cadetia.simplicadet.R;
import com.cadetia.simplicadet.database.DbQuery;
import com.cadetia.simplicadet.listeners.MyCompleteListener;
import com.google.android.material.progressindicator.CircularProgressIndicator;

import java.util.ArrayList;
import java.util.List;

import eightbitlab.com.blurview.BlurView;

public class PathSelector extends AppCompatActivity {

    private SharedPreferences prefs;
    private LinearLayout pathView;
    private BlurView blurView;
    private View topBarBlurBackground;
    private ScrollView scrollView;
    private List<DbQuery.LearningPath> paths = new ArrayList<>();
    private boolean loading = true;
    private String error = null;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_path_selector);

        prefs = getSharedPreferences("LearningPathProgress", Context.MODE_PRIVATE);

        initializeViews();
        setupBlurView();
        setupScrollListener();
        setupBackButton();
        loadPaths();
    }

    private void initializeViews() {
        pathView = findViewById(R.id.pathView);
        blurView = findViewById(R.id.blur_view_path);
        topBarBlurBackground = findViewById(R.id.path_blur_background);
        scrollView = findViewById(R.id.path_scroll_view);
    }

    private void setupBlurView() {
        View decorView = getWindow().getDecorView();
        ViewGroup rootView = decorView.findViewById(android.R.id.content);
        Drawable windowBackground = decorView.getBackground();

        blurView.setupWith(rootView)
                .setFrameClearDrawable(windowBackground)
                .setBlurRadius(10f)
                .setBlurAutoUpdate(true);
    }

    private void setupScrollListener() {
        scrollView.setOnScrollChangeListener((v, scrollX, scrollY, oldScrollX, oldScrollY) -> {
            float threshold = 50f;
            float alpha = Math.min(1f, scrollY / threshold);
            topBarBlurBackground.setAlpha(alpha);
        });
    }

    private void setupBackButton() {
        ImageButton backButton = findViewById(R.id.community_back);
        backButton.setOnClickListener(v -> finish());
    }

    private void loadPaths() {

        showLoading();

        try {
            DbQuery.loadAllLearningPaths(new MyCompleteListener() {
                @Override
                public void onSucces() {
                    List<DbQuery.LearningPath> realPaths = DbQuery.g_allLearningPaths != null ?
                            DbQuery.g_allLearningPaths : new ArrayList<>();

                    paths.clear();
                    paths.addAll(realPaths);

                    loading = false;
                    error = null;

                    new Handler(Looper.getMainLooper()).post(() -> updateUI());
                }

                @Override
                public void onFailure() {
                    paths.clear();
                    loading = false;
                    error = "Failed to load";

                    new Handler(Looper.getMainLooper()).post(() -> updateUI());
                }
            });
        } catch (Exception e) {
            Log.e("PathSelector", "load error", e);
            paths.clear();
            loading = false;
            error = "Error: " + e.getMessage();

            new Handler(Looper.getMainLooper()).post(() -> updateUI());
        }
    }

    private void showLoading() {
        pathView.removeAllViews();

        // Create loading indicator
        LinearLayout loadingContainer = new LinearLayout(this);
        loadingContainer.setOrientation(LinearLayout.VERTICAL);
        loadingContainer.setGravity(android.view.Gravity.CENTER);
        loadingContainer.setLayoutParams(new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                (int) (200 * getResources().getDisplayMetrics().density)
        ));

        CircularProgressIndicator progressIndicator = new CircularProgressIndicator(this);
        progressIndicator.setLayoutParams(new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
        ));

        loadingContainer.addView(progressIndicator);
        pathView.addView(loadingContainer);
    }

    private void updateUI() {
        pathView.removeAllViews();

        if (loading) {
            showLoading();
        } else if (error != null) {
            showError();
        } else if (paths.isEmpty()) {
            showEmptyState();
        } else {
            showPaths();
        }
    }

    private void showError() {
        LinearLayout errorContainer = new LinearLayout(this);
        errorContainer.setOrientation(LinearLayout.VERTICAL);
        errorContainer.setGravity(android.view.Gravity.CENTER);
        errorContainer.setLayoutParams(new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                (int) (200 * getResources().getDisplayMetrics().density)
        ));

        TextView errorText = new TextView(this);
        errorText.setText("Error: " + error);
        errorText.setTextAlignment(View.TEXT_ALIGNMENT_CENTER);
        errorText.setTextColor(getColor(android.R.color.holo_red_dark));

        // Add some margin
        LinearLayout.LayoutParams textParams = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
        );
        textParams.setMargins(0, 0, 0, (int) (16 * getResources().getDisplayMetrics().density));
        errorText.setLayoutParams(textParams);

        // Retry button
        TextView retryButton = new TextView(this);
        retryButton.setText("Retry");
        retryButton.setBackgroundResource(android.R.drawable.btn_default);
        retryButton.setPadding(
                (int) (16 * getResources().getDisplayMetrics().density),
                (int) (8 * getResources().getDisplayMetrics().density),
                (int) (16 * getResources().getDisplayMetrics().density),
                (int) (8 * getResources().getDisplayMetrics().density)
        );
        retryButton.setOnClickListener(v -> {
            loading = true;
            error = null;
            loadPaths();
        });

        errorContainer.addView(errorText);
        errorContainer.addView(retryButton);
        pathView.addView(errorContainer);
    }

    private void showEmptyState() {
        LinearLayout emptyContainer = new LinearLayout(this);
        emptyContainer.setOrientation(LinearLayout.VERTICAL);
        emptyContainer.setGravity(android.view.Gravity.CENTER);
        emptyContainer.setLayoutParams(new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                (int) (200 * getResources().getDisplayMetrics().density)
        ));

        TextView emptyText = new TextView(this);
        emptyText.setText("No learning paths");
        emptyText.setTextAlignment(View.TEXT_ALIGNMENT_CENTER);

        emptyContainer.addView(emptyText);
        pathView.addView(emptyContainer);
    }

    private void showPaths() {
        pathView.removeAllViews();
        pathView.setOrientation(LinearLayout.VERTICAL);

        for (DbQuery.LearningPath path : paths) {
            View pathCard = createLearningPathCard(path);
            pathView.addView(pathCard);

            View spacer = new View(this);
            LinearLayout.LayoutParams spacerParams = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    (int) (5 * getResources().getDisplayMetrics().density)
            );
            spacer.setLayoutParams(spacerParams);
            pathView.addView(spacer);
        }
    }

    private View createLearningPathCard(DbQuery.LearningPath path) {
        LayoutInflater inflater = LayoutInflater.from(this);
        LinearLayout cardView = (LinearLayout) inflater.inflate(R.layout.item_learning_path_header, pathView, false);

        // Set click listener
        cardView.setOnClickListener(v -> selectPath(path));

        // Find views
        TextView titleView = cardView.findViewById(R.id.pathTitle);
        TextView lessonsView = cardView.findViewById(R.id.pathLessons);
        CircularProgressIndicator progressView = cardView.findViewById(R.id.circularProgress);
        TextView progressText = cardView.findViewById(R.id.progressText);

        // Calculate progress
        int progress = calculateProgress(path);
        int total = path.nodes != null ? path.nodes.size() : 0;

        // Set data
        titleView.setText(path.title != null ? path.title : "Unknown");
        lessonsView.setVisibility(View.VISIBLE);
        lessonsView.setText(total + " lessons");
        progressView.setProgress(progress);
        progressText.setText(progress + "%");

        return cardView;
    }

    private int calculateProgress(DbQuery.LearningPath path) {
        if (path.nodes == null || path.nodes.isEmpty()) {
            return 0;
        }

        int completedCount = 0;
        for (DbQuery.LearningPathNode node : path.nodes) {
            if (prefs.getBoolean(path.id + "_" + node.id, false)) {
                completedCount++;
            }
        }

        return (completedCount * 100) / path.nodes.size();
    }

    private void selectPath(DbQuery.LearningPath path) {
        SharedPreferences selectedPathPrefs = getSharedPreferences("SelectedLearningPath", Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = selectedPathPrefs.edit();
        editor.putString("selectedPathId", path.id);
        editor.putString("selectedPathTitle", path.title);
        editor.apply();

        DbQuery.g_learningPath = path;
        finish();
    }
}

File: simplicadet.app/app/src/main/java/com/cadetia/simplicadet/activities/PathSelector.kt
package com.cadetia.simplicadet.activities

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.cadetia.simplicadet.activities.ui.theme.SimplicadetappTheme
import com.cadetia.simplicadet.database.DbQuery
import com.cadetia.simplicadet.listeners.MyCompleteListener

class PathSelector : ComponentActivity() {
    private lateinit var learningPathPrefs: SharedPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()

        learningPathPrefs = getSharedPreferences("LearningPathProgress", Context.MODE_PRIVATE)

        setContent {
            SimplicadetappTheme {
                Surface(
                    modifier = Modifier.fillMaxSize(),
                    color = MaterialTheme.colorScheme.background
                ) {
                    PathSelectorScreen(
                        onPathSelected = { learningPath ->
                            selectLearningPath(learningPath)
                        },
                        learningPathPrefs = learningPathPrefs
                    )
                }
            }
        }
    }

    private fun selectLearningPath(learningPath: DbQuery.LearningPath) {
        // Save selected learning path to preferences
        val editor = getSharedPreferences("SelectedLearningPath", Context.MODE_PRIVATE).edit()
        editor.putString("selectedPathId", learningPath.id)
        editor.putString("selectedPathTitle", learningPath.title)
        editor.apply()

        // Set as current learning path in DbQuery
        DbQuery.g_learningPath = learningPath

        // Go back to home
        finish()
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun PathSelectorScreen(
    onPathSelected: (DbQuery.LearningPath) -> Unit,
    learningPathPrefs: SharedPreferences
) {
    val context = LocalContext.current
    var learningPaths by remember { mutableStateOf<List<DbQuery.LearningPath>>(emptyList()) }
    var isLoading by remember { mutableStateOf(true) }
    var error by remember { mutableStateOf<String?>(null) }

    // Load all learning paths
    LaunchedEffect(Unit) {
        loadAllLearningPaths { paths, errorMsg ->
            learningPaths = paths
            error = errorMsg
            isLoading = false
        }
    }

    Scaffold(
        topBar = {
            TopAppBar(
                title = {
                    Text(
                        "Select Learning Path",
                        fontSize = 20.sp,
                        fontWeight = FontWeight.Bold
                    )
                },
                navigationIcon = {
                    IconButton(onClick = {
                        (context as? ComponentActivity)?.finish()
                    }) {
                        Text("←", fontSize = 24.sp)
                    }
                }
            )
        }
    ) { paddingValues ->
        Box(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues)
        ) {
            when {
                isLoading -> {
                    CircularProgressIndicator(
                        modifier = Modifier.align(Alignment.Center)
                    )
                }
                error != null -> {
                    Column(
                        modifier = Modifier.align(Alignment.Center),
                        horizontalAlignment = Alignment.CenterHorizontally
                    ) {
                        Text(
                            text = "Error: $error",
                            color = Color.Red,
                            textAlign = TextAlign.Center
                        )
                        Spacer(modifier = Modifier.height(16.dp))
                        Button(
                            onClick = {
                                isLoading = true
                                error = null
                                loadAllLearningPaths { paths, errorMsg ->
                                    learningPaths = paths
                                    error = errorMsg
                                    isLoading = false
                                }
                            }
                        ) {
                            Text("Retry")
                        }
                    }
                }
                learningPaths.isEmpty() -> {
                    Text(
                        text = "No learning paths available",
                        modifier = Modifier.align(Alignment.Center),
                        textAlign = TextAlign.Center
                    )
                }
                else -> {
                    LazyColumn(
                        modifier = Modifier
                            .fillMaxSize()
                            .padding(16.dp),
                        verticalArrangement = Arrangement.spacedBy(12.dp)
                    ) {
                        items(learningPaths) { learningPath ->
                            LearningPathCard(
                                learningPath = learningPath,
                                learningPathPrefs = learningPathPrefs,
                                onPathClick = { onPathSelected(learningPath) }
                            )
                        }
                    }
                }
            }
        }
    }
}

@Composable
fun LearningPathCard(
    learningPath: DbQuery.LearningPath,
    learningPathPrefs: SharedPreferences,
    onPathClick: () -> Unit
) {
    // Calculate progress
    val progress = remember(learningPath) {
        if (learningPath.nodes.isNullOrEmpty()) {
            0
        } else {
            val completedNodes = learningPath.nodes.count { node ->
                val nodeKey = "${learningPath.id}_${node.id}"
                learningPathPrefs.getBoolean(nodeKey, false)
            }
            (completedNodes * 100) / learningPath.nodes.size
        }
    }

    val totalNodes = learningPath.nodes?.size ?: 0

    Card(
        modifier = Modifier
            .fillMaxWidth()
            .clickable { onPathClick() },
        elevation = CardDefaults.cardElevation(defaultElevation = 4.dp),
        shape = RoundedCornerShape(12.dp)
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp)
        ) {
            // Header with title and progress
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = learningPath.title ?: "Unknown Path",
                    fontSize = 18.sp,
                    fontWeight = FontWeight.Bold,
                    modifier = Modifier.weight(1f)
                )

                // Progress circle
                Box(
                    contentAlignment = Alignment.Center,
                    modifier = Modifier
                        .size(60.dp)
                        .clip(RoundedCornerShape(30.dp))
                        .background(
                            if (progress == 100) Color.Green.copy(alpha = 0.1f)
                            else Color.Blue.copy(alpha = 0.1f)
                        )
                ) {
                    Text(
                        text = "$progress%",
                        fontSize = 14.sp,
                        fontWeight = FontWeight.Bold,
                        color = if (progress == 100) Color.Green else Color.Blue
                    )
                }
            }

            Spacer(modifier = Modifier.height(8.dp))

            // Progress bar
            LinearProgressIndicator(
                progress = progress / 100f,
                modifier = Modifier
                    .fillMaxWidth()
                    .height(8.dp)
                    .clip(RoundedCornerShape(4.dp)),
                color = if (progress == 100) Color.Green else Color.Blue
            )

            Spacer(modifier = Modifier.height(8.dp))

            // Path info
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween
            ) {
                Text(
                    text = "$totalNodes lessons",
                    fontSize = 12.sp,
                    color = Color.Gray
                )

                Text(
                    text = when {
                        progress == 100 -> "Completed"
                        progress > 0 -> "In Progress"
                        else -> "Not Started"
                    },
                    fontSize = 12.sp,
                    color = when {
                        progress == 100 -> Color.Green
                        progress > 0 -> Color.Blue
                        else -> Color.Gray
                    },
                    fontWeight = FontWeight.Medium
                )
            }
        }
    }
}

// Helper function to load all learning paths
private fun loadAllLearningPaths(callback: (List<DbQuery.LearningPath>, String?) -> Unit) {
    // This would need to be implemented in your DbQuery class
    // For now, I'll show how it should work with the existing structure

    try {
        DbQuery.loadAllLearningPaths(object : MyCompleteListener {
            override fun onSucces() {
                // Assuming DbQuery.g_allLearningPaths would contain all paths
                val paths = DbQuery.g_allLearningPaths ?: emptyList()
                callback(paths, null)
            }

            override fun onFailure() {
                callback(emptyList(), "Failed to load learning paths")
            }
        })
    } catch (e: Exception) {
        Log.e("PathSelector", "Error loading learning paths", e)
        callback(emptyList(), "Error: ${e.message}")
    }
}